display-points-using-clustering-feature-reduction/src/main/java/com/esri/arcgismaps/sample/displaypointsusingclusteringfeaturereduction/components/MapViewModel.kt

```diff
@@ -0,0 +1,232 @@
+
+class MapViewModel(
+    application: Application,
+    private val sampleCoroutineScope: CoroutineScope
+) : AndroidViewModel(application) {
+    // set the MapView mutable stateflow
+    val mapViewState = MutableStateFlow(MapViewState())
+
+    // create a ViewModel to handle dialog interactions
+    val messageDialogVM: MessageDialogViewModel = MessageDialogViewModel()
+
+    // Flag indicating whether feature reduction is enabled or not
+    val isFeatureReductionEnabled = mutableStateOf(true)
+
+    // Flag to show or dismiss the LoadingDialog
+    val showLoadingDialog = mutableStateOf(false)
+
+    // Flag to show or dismiss the bottom sheet
+    val showClusterSummaryBottomSheet = mutableStateOf(false)
+
+    // Initialize clusterInfoList which holds the popup details
+    val clusterInfoList = mutableStateListOf<AnnotatedString>()
+
+    // the title of the popup result
+    val popupTitle = mutableStateOf("")
+
+    init {
+        // show loading dialog to indicate that the map is loading
+        showLoadingDialog.value = true
+        // load the portal and create a map from the portal item
+        val portalItem = PortalItem(
+            Portal(application.getString(R.string.portal_url)),
+            "8916d50c44c746c1aafae001552bad23"
+        )
+
+        // set the map to be displayed in the layout's MapView
+        mapViewState.value.arcGISMap = ArcGISMap(portalItem)
+
+        sampleCoroutineScope.launch {
+            mapViewState.value.arcGISMap.load().onSuccess {
+                showLoadingDialog.value = false
+            }
+        }
+    }
+
+    /**
+    `* Toggle the FeatureLayer's featureReduction property
+     */
+    fun toggleFeatureReduction() {
+        val map = mapViewState.value.arcGISMap
+        isFeatureReductionEnabled.value = !isFeatureReductionEnabled.value
+        if (map.loadStatus.value == LoadStatus.Loaded) {
+            map.operationalLayers.forEach { layer ->
+                when (layer) {
+                    is FeatureLayer -> {
+                        layer.featureReduction?.isEnabled = isFeatureReductionEnabled.value
+                    }
+
+                    else -> {}
+                }
+            }
+        }
+    }
+
+    /**
+     * Identify the feature layer results and display the resulting popup element information
+     */
+    fun handleIdentifyResult(result: Result<List<IdentifyLayerResult>>) {
+        sampleCoroutineScope.launch {
+            result.onSuccess { identifyResultList ->
+                // initialize the string for each tap event resulting in a new identifyResultList
+                clusterInfoList.clear()
+                popupTitle.value = ""
+                identifyResultList.forEach { identifyLayerResult ->
+                    val popups = identifyLayerResult.popups
+                    popups.forEach { popup ->
+                        // set the popup title
+                        popupTitle.value = popup.title
```

Comment:
I tested with the sample map and the map we have in example app, the popup title always remain the same which is "Cluster Summary", so think we don't need to append it each time

Reply:
That is because of the popup configuration of this webmap. Some webmaps may use attribute values or expression to define popup title. But I think it is fine for this sample.